File: iter1/307.cpp
//
// Created by chenyufan on 2020/4/21.
//

#include <vector>
#include <iostream>
#include <cmath>

using namespace std;

class NumArray {
    vector<vector<int>> tree;
    int size;

public:
    NumArray(vector<int>& nums) {
        size = nums.size();
        tree.push_back(nums);

        int l = 1;
        while (tree[l-1].size()) {
            if (tree.size() <= l) tree.push_back(vector<int>(nums.));
            int idx = 0;
            for (int i = 0; i < tree[l-1].size() / 2; ++i) {
                tree[l][idx] = tree[l-1][idx*2];
                if (idx*2 + 1 < tree[l-1].size()) tree[l][idx] += tree[l-1][idx*2+1];
            }
            l++;
        }
    }

    void update(int i, int val) {
        int l = 0, idx = i;
        do {
            tree[l][idx] += val - tree[l][idx];
            l++;
            idx /= 2;
        } while(idx > 0);
    }

    int sumRange(int i, int j) {
        // get common ancestor
        int l = 0, lidx = i, ridx = j;
        while (lidx != ridx) {
            l++;
            lidx /= 2;
            ridx /= 2;
        }
        int res = tree[l][lidx];
        if (i%2 && i-1>=0) {
            res -= tree[0][i-1];
        }
        if (j%2 == 0 && j+1 < tree[0].size()) {
            res -= tree[0][j+1];
        }
        return res;
    }
};

/**
 * Your NumArray object will be instantiated and called as such:
 * NumArray* obj = new NumArray(nums);
 * obj->update(i,val);
 * int param_2 = obj->sumRange(i,j);
 */

int main() {
    vector<int> v{1,3,5};
    NumArray *obj = new NumArray(v);
    cout << obj->sumRange(0, 2) << endl;
    obj->update(1, 2);
    cout << obj->sumRange(0, 2) << endl;
}

File: iter2/307.py
# 线段树

from dataclasses import dataclass
@dataclass
class SegmentTreeNode:
    start : int
    end : int
    mid : int
    sum : int
    left : "SegmentTreeNode"
    right : "SegmentTreeNode"

class NumArray:

    def __init__(self, nums: List[int]):
        def buildTree(lo, hi):
            if lo > hi:
                return None
            if lo == hi:
                return SegmentTreeNode(lo, lo, lo, nums[lo], None, None)
            mid = (lo + hi) // 2
            left = buildTree(lo, mid)
            right = buildTree(mid+1, hi)
            return SegmentTreeNode(lo, hi, mid, left.sum+right.sum, left, right)

        self.tree = buildTree(0, len(nums)-1)

    def update(self, i: int, val: int) -> None:
        def _update(node):
            if node.start == node.end == i:
                node.sum = val
                return
            _update(node.left) if node.mid >= i else _update(node.right)
            node.sum = node.left.sum + node.right.sum
        _update(self.tree)


    def sumRange(self, i: int, j: int) -> int:
        def _query(node, lo, hi):
            if node.start == lo and node.end == hi:
                return node.sum

            if lo <= hi <= node.mid:
                return _query(node.left, lo, hi)
            elif hi >= lo >= node.mid+1:
                return _query(node.right, lo, hi)
            else:
                return _query(node.left, lo, node.mid) + _query(node.right, node.mid+1, hi)
        return _query(self.tree, i, j)


# Your NumArray object will be instantiated and called as such:
# obj = NumArray(nums)
# obj.update(i,val)
# param_2 = obj.sumRange(i,j)

